Support "nf_conntrack", check for 90% full

arekkusu opened the pull request:
- Script was checking for "ip_conntrack_..." which was replaced by "nf_conntrack_..." on newer system. Now support both.

-  Return failure ("not ok") when table is more than 90% full. Note sure what value is best here but I think that is better than when the table is full. Otherwise we might end up with a value close to the max or bouncing around.

- Replaced cat by "$(< file )" to avoid calling external command

Random-Liu commented:
/ok-to-test

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
 
 # This plugin checks for common network issues. Currently, it only checks
-# if the conntrack table is full. 
+# if the conntrack table is full.
```
We should also change this comment:
"# if the conntrack table is more than 90% full."

arekkusu replied:
Thanks, update the comment.

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
 
-if (( conntrack_count >= conntrack_max )); then
-    echo "Conntrack table full"
+if (( conntrack_count > conntrack_max /10 * 9)); then
```
nit: Bash only does integer arithmetic. To avoid/reduce precision loss problems, it's best to change this to ` *9 /10`.

arekkusu replied:
Very good point ! Updated as proposed. Thanks.

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
 
-conntrack_max=$(cat /proc/sys/net/ipv4/netfilter/ip_conntrack_max)
-conntrack_count=$(cat /proc/sys/net/ipv4/netfilter/ip_conntrack_count)
+if [ -f $NF_CONNTRACK_COUNT ] && [ -f $NF_CONNTRACK_MAX ]; then
```
nit: Could you help change the `[ -f xxx ]`s here and below into `[[ -f xxx ]]`? Thanks! :)
This is because `[[ ]]` is [more prefered](https://google.github.io/styleguide/shell.xml#Test,_%5B_and_%5B%5B) than `[ ]`. If you could help use this chance to help improve the code quality, I'm sure everyone would appreciate it :)

arekkusu replied:
Updated to use `[[ ]]`, I also switched to 2 space indent to follow Google Shell Style Guide.

I think same should be done for `check_ntp.sh`.

xueweiz replied:
Thanks! Yes I agree, we should update `check_ntp.sh` as well as a small improvement.

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
-[ -f /proc/sys/net/ipv4/netfilter/ip_conntrack_max ] || exit $UNKNOWN
-[ -f /proc/sys/net/ipv4/netfilter/ip_conntrack_count ] || exit $UNKNOWN
+# "nf_conntrack" replaces "ip_conntrack" - support both
+NF_CONNTRACK_COUNT='/proc/sys/net/netfilter/nf_conntrack_count'
```
nit: Could you add `readonly ` before the [constants](https://google.github.io/styleguide/shell.xml#Constants_and_Environment_Variable_Names) (`NF_CONNTRACK_COUNT`, `NF_CONNTRACK_MAX`...) and the [read-only variables](https://google.github.io/styleguide/shell.xml#Read-only_Variables) (`conntrack_count`, `conntrack_max`)?

Thank you!

arekkusu replied:
I added `readonly` where necessary. 
Feels like a lot of `readonly` for a small script (readability wise) but I understand the reasoning :)

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
+    exit $UNKNOWN
+fi
+
+conntrack_count=$(< $CONNTRACK_COUNT) || exit $UNKNOWN
```
nit: We have `conntrack_count` and `CONNTRACK_COUNT`, one for the value and another one for the path. do you mind appending `_PATH` to the path constants? Thanks so much!

arekkusu replied:
Not a nit, you're right about the naming.
I added `PATH` and shortened `CONNTRACK` to `CT`

xueweiz replied:
Sounds good to me! Thanks! :)

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
-if (( conntrack_count >= conntrack_max )); then
-    echo "Conntrack table full"
+if (( conntrack_count > conntrack_max /10 * 9)); then
+    echo "Conntrack table is more than 90% full"
```
Maybe
`echo "Conntrack table is more than 90% used: ${conntrack_count} out of ${conntrack_max}"`

What do you think?

arekkusu replied:
Update to included this additional information for both `OK` and `NONOK` exit.
In addition I choose to output to `STDERR` for `NONOK`

xueweiz replied:
Sorry for the trouble. Could you help change the print back to `STDOUT`? :P

This is a small detail in `custom-plugin-monitor`'s interface, where only `STDOUT` is collected, and `STDERR` is ignored. See the [design](https://docs.google.com/document/d/1jK_5YloSYtboj-DtfjmYKxfNnUxCAvohLnsH5aGCAYQ/edit#) and the [implementation](https://github.com/kubernetes/node-problem-detector/blob/8704ec0c42b27db5eca8e57c3dbfc172ba82765d/pkg/custompluginmonitor/plugin/plugin.go#L131) for this detail.

xueweiz commented on config/plugin/network_problem.sh:
```diff
@@ -1,23 +1,35 @@
+    exit $UNKNOWN
+fi
+
+conntrack_count=$(< $CONNTRACK_COUNT) || exit $UNKNOWN
```
Thanks for replacing `cat` with `< ` :)

arekkusu replied:
:)

arekkusu commented:
Really appreciate your review and input. Have a look at the update and let me know what you think.

xueweiz commented:
Hi @arekkusu , thanks a lot for making the changes! Just one more small item for the `STDOUT` thing :) once that's fixed I'll approve and merge the PR.
Thanks!

xueweiz commented:
And by the way, you can (optionally) squash all the 2-3 commits into one. We typically do this to have a simpler/cleaner commit history. But I understand that a lot of people choose to not do that to keep a history for the code review.
Your choice :) I'm fine with either way.

arekkusu commented:
Ok regarding outputting to `STDOUT`. Updated and squashed the commits.

BTW do see any changes needed to network-problem-monitor.json`?
I guess `ConntrackFull` is no longer fully accurate ( `ConntrackPressure` or `ConntrackHighUsage`). I am not sure it's something we want / need to change.

https://github.com/kubernetes/node-problem-detector/blob/599ca532e81833debb4791c0410464507a80c869/config/network-problem-monitor.json#L1-L20

arekkusu commented:
I think the (draft?) design doc doesn't clearly indicate the UNKNOWN return status.

Implementation considers anything else than 0 or 1 as unknown

https://github.com/kubernetes/node-problem-detector/blob/8704ec0c42b27db5eca8e57c3dbfc172ba82765d/pkg/custompluginmonitor/plugin/plugin.go#L153-L161

This test does refer to exit code 2 as unkown

https://github.com/kubernetes/node-problem-detector/blob/8704ec0c42b27db5eca8e57c3dbfc172ba82765d/pkg/custompluginmonitor/plugin/test-data/unknown.sh#L1-L5

Let me know if you think it's better to use 3 as exit code for UNKNOWN ...

xueweiz commented:
> I guess `ConntrackFull` is no longer fully accurate (`ConntrackPressure` or `ConntrackHighUsage`). I am not sure it's something we want / need to change.

I agree with you, I think `ConntrackPressure` seems like a good option. However I don't know whether it's worth it to change it...Since users may have already doing detecting & remedy based on the `ConntrackFull` event that NPD reports. Changing it is more visible to the users. I don't have a strong opinion in either way. I think your PR is fine as is.
+cc @andyxning , who reviewed the original PR #152 .

> Let me know if you think it's better to use 3 as exit code for UNKNOWN ...

I think keeping it as is (which is `2`) is OK :)
(I'm quite curious why was this not clearly defined lol

/lgtm
/approve
